lib/pure/os.nim

```diff
@@ -2088,6 +2088,216 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
               k = getSymlinkFileKind(path)
             yield (k, y)
 
+type OpenDirStatus* = enum
+  odOpenOk,
+  odNotDir,
+  odAccessDenied,
+  odNotFound,
+  odUnknownError
+
+type WalkStepKind* = enum
+  wsOpenDir,     ## attempt to open directory
+  wsEntryOk,     ## the entry is OK
+  wsEntryBad     ## a broken symlink on posix, where it's unclear if it
+                 ## points to a file or directory
+  wsInterrupted  ## the directory handle got invalidated during reading
+
+type WalkStep = ref object
+  path*: string
+  code*: OSErrorCode
+  case kind*: WalkStepKind
+  of wsOpenDir:
+    openStatus*: OpenDirStatus
+  of wsEntryOk:
+    entryType*: PathComponent
+  of wsEntryBad, wsInterrupted:
+    discard
+
+proc `$`*(s: WalkStep): string =
+  let beg =
+    case s.kind
+    of wsOpenDir: "(wsOpenDir | " & $s.openStatus
+    of wsEntryOk: "(wsEntryOk | '" & s.path & "', " & $s.entryType
+    of wsEntryBad: "(wsEntryBad | '" & s.path & "'"
+    of wsInterrupted: "(wsInterrupted | '" & s.path & "'"
+  result = beg & ", err:" & $s.code & ")"
+
+iterator tryWalkDir*(dir: string; relative=false): WalkStep {.
+  tags: [ReadDirEffect], raises: [], since: (1,1), noNimScript.} =
+  ## A version of `walkDir iterator <#walkDir.i,string>`_ with more
+  ## thorough error checking. It never raises an exception.
+  ## Yields *steps* of walking through `dir`, each step contains
+  ## its path ``path``, OS error code ``code`` and additional tag ``kind``.
+  ## At first step it yields info ``openStatus`` of opening
+  ## status `dir` as a directory, tag ``wsOpenDir``.
+  ## If it's OK, the subsequent steps will yield the directory entries
+  ## with component type as info ``entryType``, tag ``wsEntryOk``.
+  ## Dangling symlinks on posix are reported as ``wsEntryBad``.
+  ## If (rarely) walking terminates with an error (such as an I/O error),
+  ## tag is `wsInterrupted`.
```

Comment:
Please reword this to not be a wall of text.

Split into paragraphs:

1. Short description of what this is.
2. Longer description of how it works.
3. Edge cases/other details.

Reply:
I tried to improve that.